src/natural_history_manager.rs

```diff
@@ -0,0 +1,572 @@
+        self.set_person_property(person_id, NaturalHistoryIdx, Some(index));
+    }
+
+    fn evaluate_inverse_generation_interval(
+        &self,
+        person_id: PersonId,
+        gi_cdf_value_unif: f64,
+    ) -> f64 {
+        // Let's first deal with the corner case -- the person is experiencing their first infection attempt.
+        // In this case, gi_cdf_value_unif will be 0.0. There are no points below 0.0 in a CDF, so interpolation
+        // will fail. Instead, we return 0.0. This is the obvious value because it means that the person
+        // has experienced none of their infectiousness at the start of their infection. It also ensures that if
+        // GI CDF is 0.0 for some time after the start of infection, inverse_gi(\epsilon) - inverse_gi(0) = c > 0
+        // even as \epsilon -> 0, which properly reproduces a CDF where an individual is not infectious immediately.
+        if gi_cdf_value_unif == 0.0 {
+            return 0.0;
+        }
+        // Grab the set GI trajectory for this person.
+        let gi_index = self
+            .get_person_property(person_id, NaturalHistoryIdx)
+            .expect("No GI index set. Has this person been infected?");
+        let natural_history_container = self
+            .get_data_container(NaturalHistory)
+            .expect("Natural history data container not initialized.");
+        let times = &natural_history_container.times[gi_index];
+        let gi_trajectory = &natural_history_container.gi_trajectories[gi_index];
+        // Because we want to interpolate the *inverse* CDF, the CDF values are "x" and the time values are "y".
+        interpolate(gi_trajectory, times, gi_cdf_value_unif)
+    }
+}
+
+/// An interpolation routine that expects a paired set of values `xs` and `ys` that represent samples
+/// from a given function. The function is evaluated at a given x value `xp` using cubic spline interpolation
+/// when there are at least two samples above and below `xp`. Otherwise, it uses linear extrapolation at the tails.
+/// Assumes that function samples are sorted so that the `xs` are in ascending order.
+fn interpolate(xs: &[f64], ys: &[f64], xp: f64) -> f64 {
+    let upper_window_index_option = xs.iter().position(|&x| x > xp);
+    // We need to check whether a point was found. If it wasn't, it means that all values
+    // in `xs` are less than `xp`. We have to use an alternative extrapolation strategy.
+    let upper_window_index = match upper_window_index_option {
+        None => {
+            // We are above the range of the `xs` samples, so we must extrapolate. We default to
+            // linear extrapolation using the last two values in `xs`.
+            let traj_len = xs.len();
+            return linear_interpolation(
+                xs[traj_len - 2],
+                xs[traj_len - 1],
+                ys[traj_len - 2],
+                ys[traj_len - 1],
+                xp,
+            );
+        }
```

Comment:
What happens if someone provides `xp` really far in the future? Won't you extrapolate above 1?

Reply:
In this case, we're extrapolating the inverse CDF, so the $x$ values are the values of the generation interval CDF and the $y$ values are time. We can't extrapolate above 1, but you are correct that we can extrapolate some time that is very far away in the future as $xp \to 1$. But, that's also the behavior we expect from the tail of a CDF.

I realize now that I have an implicit assumption here: if the CDF were bounded -- meaning there were some maximum time value possible -- the input CSV would contain that maximum CDF value and time (i.e., a point that takes the form CDF value: 1.0000, time: 42.0). I think this is a fair assumption, but it's worth stating it. I'll add this to the markdown description as well. Thanks!